values/dict.go

```diff
@@ -0,0 +1,312 @@
+	// Dictionary with the key/value removed.
+	Remove(key Value) Dictionary
+
+	// Range will iterate over each element in
+	// the Dictionary.
+	Range(func(key, value Value))
+
+	// Len returns the number of elements inside of
+	// Dictionary.
+	Len() int
+}
+
+type dict struct {
+	t    semantic.MonoType
+	data *immutable.SortedMap
+}
+
+func (d dict) Get(key, def Value) Value {
+	if !key.IsNull() {
+		v, ok := d.data.Get(key)
+		if ok {
+			return v.(Value)
+		}
+	}
+	return def
+}
+
+func (d dict) Insert(key, value Value) (Dictionary, error) {
+	if key.IsNull() {
+		return nil, errors.New(codes.Invalid, "null value cannot be used as a dictionary key")
+	}
+	data := d.data.Set(key, value)
+	return dict{data: data}, nil
+}
+
+func (d dict) Remove(key Value) Dictionary {
+	if key.IsNull() {
+		return d
+	}
+	data := d.data.Delete(key)
+	return dict{data: data}
+}
+
+func (d dict) Range(f func(key, value Value)) {
+	itr := d.data.Iterator()
+	for {
+		key, value := itr.Next()
+		if key == nil {
+			return
+		}
+		f(key.(Value), value.(Value))
+	}
+}
+
+func (d dict) Len() int {
+	return d.data.Len()
+}
+
+func (d dict) Type() semantic.MonoType { return d.t }
+func (d dict) IsNull() bool            { return false }
+func (d dict) Str() string {
+	panic(UnexpectedKind(semantic.Dictionary, semantic.String))
+}
+func (d dict) Bytes() []byte {
+	panic(UnexpectedKind(semantic.Dictionary, semantic.Bytes))
+}
+func (d dict) Int() int64 {
+	panic(UnexpectedKind(semantic.Dictionary, semantic.Int))
+}
+func (d dict) UInt() uint64 {
+	panic(UnexpectedKind(semantic.Dictionary, semantic.UInt))
+}
+func (d dict) Float() float64 {
+	panic(UnexpectedKind(semantic.Dictionary, semantic.Float))
+}
+func (d dict) Bool() bool {
+	panic(UnexpectedKind(semantic.Dictionary, semantic.Bool))
+}
+func (d dict) Time() Time {
+	panic(UnexpectedKind(semantic.Dictionary, semantic.Time))
+}
+func (d dict) Duration() Duration {
+	panic(UnexpectedKind(semantic.Dictionary, semantic.Duration))
+}
+func (d dict) Regexp() *regexp.Regexp {
+	panic(UnexpectedKind(semantic.Dictionary, semantic.Regexp))
+}
+func (d dict) Array() Array {
+	panic(UnexpectedKind(semantic.Dictionary, semantic.Array))
+}
+func (d dict) Object() Object {
+	panic(UnexpectedKind(semantic.Dictionary, semantic.Object))
+}
+func (d dict) Function() Function {
+	panic(UnexpectedKind(semantic.Dictionary, semantic.Function))
+}
+func (d dict) Dict() Dictionary {
+	return d
+}
+
+func (d dict) Equal(v Value) bool {
+	if !d.t.Equal(v.Type()) {
+		return false
+	}
+
+	other := v.Dict()
+	if d.data.Len() != other.Len() {
+		return false
+	}
+
+	equal := true
+	other.Range(func(key, value Value) {
+		if !equal {
+			return
+		}
+
+		v, ok := d.data.Get(key)
+		if !ok {
+			equal = false
+			return
+		}
+		equal = value.Equal(v.(Value))
+	})
+	return equal
+}
+
+type (
+	intComparer      struct{}
+	uintComparer     struct{}
+	floatComparer    struct{}
+	stringComparer   struct{}
+	durationComparer struct{}
```

Comment:
I thought we discovered that durations aren't comparable?

Reply:
I think I had already added it since it is possible to compare them. The sorting won't necessarily be meaningful though. I'll remove this section so there isn't unused code.